Crashlytics/Crashlytics/Models/FIRCLSInstallIdentifierModel.m

```diff
@@ -96,20 +100,33 @@ - (NSString *)generateInstallationUUID {
 
 #pragma mark Privacy Shield
 
-/**
- * To support privacy shield we need to regenerate the install id when the IID changes.
- *
- * This is a blocking, slow call that must be called on a background thread.
- */
-- (void)regenerateInstallIDIfNeededWithBlock:(void (^)(BOOL didRotate))callback {
-  // This callback is on the main thread
-  [self.installations
-      installationIDWithCompletion:^(NSString *_Nullable currentIID, NSError *_Nullable error) {
-        dispatch_async(dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_HIGH, 0), ^{
-          BOOL didRotate = [self rotateCrashlyticsInstallUUIDWithIID:currentIID error:error];
-          callback(didRotate);
-        });
-      }];
+- (BOOL)regenerateInstallIDIfNeeded {
+  BOOL __block didRotate = false;
+
+  // Run this only once because it can be run multiple times in succession,
+  // and if it's slow it could delay crash upload too much without providing
+  // user benefit.
+  dispatch_once(&_regenerateOnceToken, ^{
+    dispatch_semaphore_t semaphore = dispatch_semaphore_create(0);
+
+    // This runs Completion async, so wait a reasonable amount of time for it to finish.
+    [self.installations
+        installationIDWithCompletion:^(NSString *_Nullable currentIID, NSError *_Nullable error) {
+          didRotate = [self rotateCrashlyticsInstallUUIDWithIID:currentIID error:error];
+
+          if (didRotate) {
+            FIRCLSInfoLog(@"Rotated Crashlytics Install UUID because Firebase Install ID changed.");
+          }
+          dispatch_semaphore_signal(semaphore);
+        }];
+
+    intptr_t result = dispatch_semaphore_wait(
```

Comment:
Just want to double check that this part is never called and never blocks main thread as it will trigger a deadlock (at least for FIRCLSInstallationsWaitTime) because installationIDWithCompletion completion handler is called on the main thread.

Reply:
I guess I can just put this on a background thread to make sure that's the case

Reply:
Alright we call this from a background queue so we should be good
![Screen Shot 2021-04-21 at 11 55 53 AM](https://user-images.githubusercontent.com/555046/115584146-8b41b500-a298-11eb-86c2-9d043c104df3.png)